### DISCUSSION TITLE
Default Nullable for type or not?

### DISCUSSION THREAD

**Discussion by wangzhen11aaa** [databendlabs/databend#4234] [General]
### Context
  I am doing the task that supports the nullable group-by, running into some bewilderments.  Our database makes the type nullable default, quote 
`create table t(a UInt64, b UInt32) Engine = Fuse;
insert into t(a,b)  select if (number % 3 = 1, null, number) as a, number + 3 as b from numbers(10);
` from https://github.com/datafuselabs/databend/issues/4079
When we do the following query search like `select a from t group by a`,we will use 8 bytes + 1 byte size to store this hash_key. I mean from this simple query, we can infer that all the types will be nullable and cost 1 byte more in the hash key when used as a group by key.  Actually, we may not have any non-nullable type any more, unless that we have the type to explictly declare it as nonnullable(type), we lost some flexibility.

### The other database design (mainly clickhosue):
1. We know that the nullable is explictly declared in _Clickhouse_, It will makes it more economy in space while lose some flexibility.

### For me
I just think that  we lost some basic non-nullable value type may be not symmetric and it will drag all the hash_methods into the nullable desert. (serielize and deserilize with all the following bitmap). 

### Quote from `sundyli`
_doris，mysql，snowflake_ : nullable default,
_clickhouse, firebolt._ non-nullable  default.

So I launched this disscussion to get more thought into it. Thank you.

**Comment by sundy-li** [comment]
For now on:

SqlType eg: `c Int8` is mapping to datavalue's `Nullable(Int8Type)`,  `c Int8 not null` is mapping to `Int8Type`.

This may be a little confusing.

So I would vote for defaults to be non-nullable data_type, which is the style of ClickHouse & firebolt, then SqlType is compatible with datavalue's datatype.

What's your opinion? cc @zhang2014  @BohuTANG  @Xuanwo

**Reply by Xuanwo** [reply]
I prefer non-nullable data_type personally (without technical consideration)

**Comment by bohutang** [comment]
I think this is a user experience problem, not a technical problem.
For most user, they don't care what the NULL is, non-nullable default data type will lower the user's barrier to use.

**Comment by sundy-li** [comment]
#4435 Defaults to not null now.

**Comment by ZeaLoVe** [comment]
I prefer nullable default.  :)

**Reply by sundy-li** [reply]
Default nullable could be more flexible when we migrate other DMBS tests.

Should we add a setting  `datatype_default_nullable` to control the behavior?  cc @leiysky  @BohuTANG  @zhang2014   @Xuanwo

**Reply by Xuanwo** [reply]
- Migrating other DMBS tests' goal is to improve the quality of databend.
- Adding new settings like `datatype_default_nullable` makes our logic and behavior more complex, maybe worse for our code quality.

From my first look, it's not a good tradeoff. I prefer to postpone this feature until we have strong requests from our *real* users.

**Reply by leiysky** [reply]
Agree with @Xuanwo
